src/main/app/index.js

```diff
@@ -231,25 +273,134 @@ class App {
     }
   }
 
-  // TODO(sessions): ...
-  // // Make Mark Text a single instance application.
-  // _makeSingleInstance() {
-  //   if (process.mas) return
-  //
-  //   app.requestSingleInstanceLock()
-  //
-  //   app.on('second-instance', (event, argv, workingDirectory) => {
-  //     // // TODO: Get active/last active window and open process arvg etc
-  //     // if (currentWindow) {
-  //     //   if (currentWindow.isMinimized()) currentWindow.restore()
-  //     //   currentWindow.focus()
-  //     // }
-  //   })
-  // }
+  _openFilesToOpen () {
+    this._openPathList(this._openFilesCache, false)
+  }
+
+  /**
+   * Open the path list in the best window(s).
+   *
+   * @param {string[]} pathsToOpen The path list to open.
+   * @param {boolean} openFilesInSameWindow Open all files in the same window with
+   * the first directory and discard other directories.
+   */
+  _openPathList (pathsToOpen, openFilesInSameWindow=false) {
+    const { _windowManager } = this
+    const { openFilesInNewWindow } = this._accessor.preferences.getAll()
+
+    const fileSet = new Set()
+    const directorySet = new Set()
+    for (const { isDir, path } of pathsToOpen) {
+      if (isDir) {
+        directorySet.add(path)
+      } else {
+        fileSet.add(path)
+      }
+    }
+
+    // Filter out directories that are already opened.
+    for (const window of _windowManager.windows.values()) {
+      if (window.type === WindowType.EDITOR) {
+        const { openedRootDirectory } = window
+        if (directorySet.has(openedRootDirectory)) {
+          window.bringToFront()
+          directorySet.delete(openedRootDirectory)
+        }
+      }
+    }
+
+    const directoriesToOpen = Array.from(directorySet).map(dir => ({ rootDirectory: dir, fileList: [] }))
+    const filesToOpen = Array.from(fileSet)
+
+    // Discard all directories except first one and add files.
+    if (openFilesInSameWindow) {
+      if (directoriesToOpen.length) {
+        directoriesToOpen[0].fileList.push(...filesToOpen)
+        directoriesToOpen.length = 1
+      } else {
+        directoriesToOpen.push({ rootDirectory: null, fileList: filesToOpen })
+      }
+      filesToOpen.length = 0
+    }
+
+    // Find the best window(s) to open the files in.
+    if (!openFilesInSameWindow && !openFilesInNewWindow) {
+      const isFirstWindow = _windowManager.getActiveEditorId() === null
+
+      // Prefer new directories
+      for (let i = 0; i < directoriesToOpen.length; ++i) {
+        const { fileList, rootDirectory } = directoriesToOpen[i]
+
+        for (let j = 0; j < filesToOpen.length; ++j) {
+          const pathname = filesToOpen[j]
+          if (isChildOfDirectory(rootDirectory, pathname)) {
+            if (isFirstWindow) {
+              fileList.push(...filesToOpen)
+              filesToOpen.length = 0
+              i = 9999
```

Comment:
What does `i = 9999` use for? Do you want to break the outer `for ... loop`? is there any better solution?

```js

for () {
  let needBreak = false
  for () {
    if () {
       needBreak = true
    }
  }
  if (needBreak) break
}
```

or set `i` to `Infinity`?